api/C/include/csi_ll_pmp.h

```diff
@@ -0,0 +1,237 @@
+ *
+ * Also, an error should be returned if the output parameters is invalid, such as idx
+ * out of range.
+ *
+ * @param idx: The index of PMP config entries
+ * @param base_addr: A pointer to the buffer which is used to output the base address
+ *                    of this entry
+ * @param size: A pointer to the buffer which is used to output the size of this entry
+ * @param attr: A point to the buffer which is used to output attribute of this entry
+ * @return : Status code
+ */
+csi_status_t csi_pmp_get_region(unsigned int idx, unsigned long *base_addr,
+                                   unsigned long *size, csi_pmp_region_attr_t *attr);
+
+/*
+ * Lock PMP entry by index.
+ *
+ * This function only can be called in M-mode. If it is called
+ * in other mode, it may return an error or cause carsh. When this function is called
+ * with success, it will not be effective untill the core has been exit M-mode.
+ *
+ * @param idx: The index of PMP config entries
+ * @return : Status code
+ */
+csi_status_t csi_pmp_lock_region(unsigned int idx);
+
+/*
+ * Set PMP entry address and size need to be controlled by index.
+ *
+ * This function only can be called in M-mode. If it is called
+ * in other mode, it may return an error or cause carsh. When this function is called
+ * with success, it will not be effective untill the core has been exit M-mode.
+ *
+ * @param idx: The index of PMP config entries
+ * @param base_addr: The base address of this entry need to be configured
+ * @param size: The size of this entry need to be configured
+ * @return : Status code
+ */
+csi_status_t csi_pmp_set_address(unsigned int idx, unsigned long base_addr,
+                                unsigned long size);
+
+/*
+ * Get PMP entry address and size been controlled by index.
+ *
+ * This function only can be called in M-mode. If it is called
+ * in other mode, it may return an error or cause carsh. When this function is called
+ * with success, it will not be effective untill the core has been exit M-mode.
+ *
+ * @param idx: The index of PMP config entries
+ * @param base_addr: A pointer to the buffer which is used to output the base address
+ *                    of this range
+ * @param size: A pointer to the buffer which is used to output the size of this entry
+ * @return : Status code
+ */
+csi_status_t csi_pmp_get_address(unsigned int idx, unsigned long *base_addr,
+                                unsigned long *size);
+
+/*
+ * Set PMP entry address mode by index.
+ *
+ * This function only can be called in M-mode. If it is called
+ * in other mode, it may return an error or cause carsh. When this function is called
+ * with success, it will not be effective untill the core has been exit M-mode.
+ *
+ * @param idx: The index of PMP config entries
+ * @param mode: The address mode of PMP config entries
+ * @return : Status code
+ */
+csi_status_t csi_pmp_set_address_mode(unsigned int idx, csi_addr_matching_t mode);
```

Comment:
The mode is bind with the address, so it should be set together with the address, and be automatically calculated. see https://github.com/riscv-software-src/opensbi/blob/master/lib/sbi/riscv_asm.c#L328

Reply:
@fanghuaqi ， I thought about it before. But there is another mode TOR. If there is no address mode input,  the driver do not know whether it should be set to TOR or other mode. Do you have any suggestion about it?